faunadb/client.py

```diff
@@ -1,14 +1,40 @@
 from time import time
 # pylint: disable=redefined-builtin
 from builtins import object
+import threading
 
 from requests import Request, Session
+from requests.auth import HTTPBasicAuth
+from requests.adapters import HTTPAdapter
 
 from faunadb.errors import _get_or_raise, FaunaError, UnexpectedError
 from faunadb.query import _wrap
 from faunadb.request_result import RequestResult
 from faunadb._json import parse_json_or_none, to_json
 
+class _Counter(object):
+  def __init__(self, init_value=0):
+    self.lock = threading.Lock()
+    self.counter = init_value
+
+  def __str__(self):
+    return "Counter(%s)" % self.counter
+
+  def increment(self):
+    with self.lock:
+      self.counter += 1
+      return self.counter
+
+  def get_and_increment(self):
```

Comment:
I understand you reasoning but, if `decrement` returns 0 means that no other client is holding this session anymore, also, at this point, if `__del__` gets called means that this client is being garbage collected, hence, no client could possibly call `new_session_client`.